core/domain/exp_services.py

```diff
@@ -2063,6 +2063,350 @@ def get_interaction_id_for_state(exp_id, state_name):
+            checkpoint was most recently reached.
+    """
+    # Fetch the model associated with the unique_progress_url_id.
+    model = exp_models.TransientCheckpointUrlModel.get(unique_progress_url_id)
+
+    # Create a model if it doesn't already exist.
+    if model is None:
+        model = exp_models.TransientCheckpointUrlModel.create(
+            exploration_id, unique_progress_url_id)
+
+    current_exploration = exp_fetchers.get_exploration_by_id(
+        exploration_id, True, exp_version)
+
+    # If the exploration is being visited the first time.
+    if model.furthest_reached_checkpoint_state_name is None:
+        model.furthest_reached_checkpoint_exp_version = exp_version
+        model.furthest_reached_checkpoint_state_name = state_name
+    elif model.furthest_reached_checkpoint_exp_version <= exp_version:
+        furthest_reached_checkpoint_exp = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                strict=True,
+                version=model.furthest_reached_checkpoint_exp_version
+            )
+        )
+        checkpoints_in_current_exp = user_services.get_checkpoints_in_order(
+            current_exploration.init_state_name, current_exploration.states)
+        checkpoints_in_older_exp = user_services.get_checkpoints_in_order(
+            furthest_reached_checkpoint_exp.init_state_name,
+            furthest_reached_checkpoint_exp.states)
+
+        # Get the furthest reached checkpoint in current exploration.
+        furthest_reached_checkpoint_in_current_exp = (
+            user_services.
+                get_most_distant_reached_checkpoint_in_current_exploration(
+                    checkpoints_in_current_exp,
+                    checkpoints_in_older_exp,
+                    model.furthest_reached_checkpoint_state_name
+                )
+        )
+
+        # If the furthest reached checkpoint doesn't exist in current
+        # exploration.
+        if furthest_reached_checkpoint_in_current_exp is None:
+            model.furthest_reached_checkpoint_exp_version = (
+                exp_version)
+            model.furthest_reached_checkpoint_state_name = state_name
+        else:
+            # Index of the furthest reached checkpoint.
+            frc_index = checkpoints_in_current_exp.index(
+                furthest_reached_checkpoint_in_current_exp)
+            # If furthest reached checkpoint is behind most recently
+            # reached checkpoint.
+            if frc_index <= checkpoints_in_current_exp.index(state_name):
+                model.furthest_reached_checkpoint_exp_version = (
+                    exp_version)
+                model.furthest_reached_checkpoint_state_name = (
+                    state_name)
+
+    model.most_recently_reached_checkpoint_exp_version = exp_version
+    model.most_recently_reached_checkpoint_state_name = state_name
+    model.creation_timestamp = datetime.datetime.utcnow()
```

Comment:
Since the timestamp is being updated let's call it `last_updated` instead of `creation_timestamp`, cause it does not really stores the creation time but the last updated time

Reply:
Done.

Reply:
Fixed.